mne/channels/channels.py

```diff
@@ -987,6 +989,118 @@ def add_channels(self, add_list, force_update_info=False):
+                                 ''.format(method))
+
+        # Check correctness of combinations
+        for this_group, these_picks in group_by.items():
+            # Check if channel indices are out of bounds
+            if not all(idx in ch_idx for idx in these_picks):
+                raise ValueError('Some channel indices are out of bounds.')
+            # Check if heterogeneous sensor type combinations
+            this_ch_type = np.array(ch_types)[these_picks]
+            if len(set(this_ch_type)) > 1:
+                types = ', '.join(set(this_ch_type))
+                raise ValueError('Cannot combine sensors of different types; '
+                                 '"{}" contains types {}.'
+                                 ''.format(this_group, types))
+            #  Check if combining less than 2 channel
+            if len(these_picks) < 2:
+                raise ValueError('Less than 2 channels in group "{}"; cannot '
+                                 'combine by method "{}".'
+                                 ''.format(this_group, method))
+            # If all good create more detailed dict with channel type
+            group_by[this_group] = dict(picks=these_picks,
+                                        ch_type=this_ch_type[0])
+
+        try:
+            # Extract stim channels if any
+            inst = self.copy().pick_types(meg=False, stim=True)
+            # Create clean info to avoid inconsistencies
+            inst.info = create_info(sfreq=self.info['sfreq'],
+                                    ch_names=inst.info['ch_names'],
+                                    ch_types=_get_channel_types(inst.info))
+        except ValueError:
+            inst = None
+
+        # Combine channels and add them to the new instance
+        for this_group, this_group_dict in group_by.items():
+            these_picks = this_group_dict['picks']
+            this_ch_type = this_group_dict['ch_type']
+            this_data = np.take(self._data, these_picks, axis=ch_axis)
+            this_group_data = np.expand_dims(method(this_data), axis=ch_axis)
+            this_info = create_info(sfreq=self.info['sfreq'],
+                                    ch_names=[this_group],
+                                    ch_types=[this_ch_type])
+            this_inst = None
+            if isinstance(self, BaseRaw):
+                this_inst = RawArray(this_group_data, this_info,
+                                     first_samp=self.first_samp)
+            elif isinstance(self, BaseEpochs):
+                this_inst = EpochsArray(this_group_data, this_info,
+                                        tmin=self.times[0])
+            elif isinstance(self, Evoked):
+                this_inst = EvokedArray(this_group_data, this_info,
+                                        tmin=self.times[0])
+            if inst is None:
+                inst = this_inst
+            else:
+                inst.add_channels([this_inst])
+
+        return inst
```

Comment:
The [Contributing to MNE webpage](https://mne.tools/stable/install/contributing.html#return-types) says that methods should return `self` and functions return copies, should I then return a modified `self` instead, @drammock?

Reply:
I think I agree with @agramfort that supporting a mix of real and virtual channels is probably a bad idea (at the very least it might cause havoc for topography plotting functions). And if I had to guess, I would say that in some cases users will want to keep the original object around (i.e., *not* modify in place), and the resulting object will at least be smaller than the original, so let's turn this into a function rather than a method.  Another consequence of that choice is that you don't have to support TFR objects right away, we can add that support later if you're not up for it right now.